### PR TITLE
add support for parsing :lang pseudo class

### PR DESCRIPTION
**Pull request by ajnirp**
this is needed for servo/servo#8219

[Review on Reviewable](https://reviewable.io/reviews/servo/rust-selectors/61)

### CODE REVIEW

**Comment by eefriedman** [comment]
(This is only a partial review; someone more familiar with the CSS parser should approve this.)



---

Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

<sup>**[src/parser.rs, line 650 [r1]](https://reviewable.io:443/reviews/servo/rust-selectors/61#-K2TcTrizShvQCHybf3d)** ([raw file](https://github.com/servo/rust-selectors/blob/df4752d1470fac317816a97ca3a858cd467602d1/src/parser.rs#L650)):</sup>
`try!(Ok(value))` is equivalent to `value`.



---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/rust-selectors/61)

File: src/parser.rs [hunk 1]
```diff
@@ -103,6 +103,7 @@ pub enum SimpleSelector {
     LastOfType,
     OnlyOfType,
     ServoNonzeroBorder,
+    Lang(String),
```
Comment (SimonSapin) [inline src/parser.rs]: Please make this an `Atom` rather than `String`, so that comparison can be fast during matching.

File: src/parser.rs [hunk 2]
```diff
@@ -642,6 +645,11 @@ fn parse_simple_pseudo_class(context: &ParserContext, name: &str) -> Result<Simp
     }
 }
 
+fn parse_lang_pseudo_class(input: &mut Parser) -> Result<SimpleSelector, ()> {
+    let string_result = Ok((try!(input.expect_ident())).into_owned());
```
Comment (SimonSapin) [inline src/parser.rs]: `Ok()` here is unnecessary if you also remove `try!()` on the next line.

File: src/tree.rs [hunk 3]
```diff
@@ -74,6 +74,9 @@ pub trait Element: Sized {
     /// Defaults to `is_link()`: when browsing history is not recorded, all links are "unvisited".
     fn is_unvisited_link(&self) -> bool { self.is_link() }
 
+    /// Returns whether this elements matches ':lang(lang)'
+    fn matches_lang(&self, lang: &str) -> bool;
```
Comment (SimonSapin) [inline src/tree.rs]: Please make the parameter an `&Atom` rather than `&str`.

**Comment by ajnirp** [comment]
Review status: 0 of 3 files reviewed at latest revision, 4 unresolved discussions.

---

<sup>**[src/parser.rs, line 106 [r1]](https://reviewable.io:443/reviews/servo/rust-selectors/61#-K2TQH1DkuRWcj_fg-oa-r1-106)** ([raw file](https://github.com/servo/rust-selectors/blob/df4752d1470fac317816a97ca3a858cd467602d1/src/parser.rs#L106)):</sup>
Done.



---

<sup>**[src/parser.rs, line 649 [r1]](https://reviewable.io:443/reviews/servo/rust-selectors/61#-K2TQH1DkuRWcj_fg-oa-r1-649)** ([raw file](https://github.com/servo/rust-selectors/blob/df4752d1470fac317816a97ca3a858cd467602d1/src/parser.rs#L649)):</sup>
Done.



---

<sup>**[src/parser.rs, line 650 [r1]](https://reviewable.io:443/reviews/servo/rust-selectors/61#-K2TcTrizShvQCHybf3d)** ([raw file](https://github.com/servo/rust-selectors/blob/df4752d1470fac317816a97ca3a858cd467602d1/src/parser.rs#L650)):</sup>
Done.



---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/rust-selectors/61)

**Comment by ajnirp** [comment]
Review status: 0 of 3 files reviewed at latest revision, 4 unresolved discussions.

---

<sup>**[src/tree.rs, line 78 [r1]](https://reviewable.io:443/reviews/servo/rust-selectors/61#-K2TQH1DkuRWcj_fg-ob-r1-78)** ([raw file](https://github.com/servo/rust-selectors/blob/df4752d1470fac317816a97ca3a858cd467602d1/src/tree.rs#L78)):</sup>
Done.



---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/rust-selectors/61)

File: src/parser.rs [hunk 4]
```diff
@@ -642,6 +645,12 @@ fn parse_simple_pseudo_class(context: &ParserContext, name: &str) -> Result<Simp
     }
 }
 
+fn parse_lang_pseudo_class(input: &mut Parser) -> Result<SimpleSelector, ()> {
+    let string = try!(input.expect_ident()).into_owned();
```
Comment (SimonSapin) [inline src/parser.rs]: Please remove `.into_owned()` here, it does an unnecessary memory allocation. (`&Cow<str>` auto-dereferences to `&str` in the `from_slice` call.)

File: src/tree.rs [hunk 5]
```diff
@@ -74,6 +74,9 @@ pub trait Element: Sized {
     /// Defaults to `is_link()`: when browsing history is not recorded, all links are "unvisited".
     fn is_unvisited_link(&self) -> bool { self.is_link() }
 
+    /// Returns whether this elements matches ':lang(lang)'
+    fn matches_lang(&self, lang: &Atom) -> bool;
```
Comment (SimonSapin) [inline src/tree.rs:78]: This is a breaking API change, please bump the version number in `Cargo.toml` to 0.3.0.

Should be all good with this!

**Comment by nox** [comment]
@SimonSapin Should `:lang()` be handled as a non-tree-structural pseudo class?

**Comment by SimonSapin** [comment]
@wenderen Sorry I forgot about this! Unfortunately GitHub does not notify me when you push to your branch to update a PR, so it’s good to also leave a comment on the PR when you do.

As @nox says, we’ve changed the parser to be generic over the set of supported pseudo-elements and "non-tree-structural" pseudo-classes. "tree-structural" means that rust-selectors can implement them itself just by traversing the element tree and does not need a callback to obtain more information about elements. This is not the case of `:lang`, so it should now be implemented in Servo rather than selectors.

But `:lang` is also a _functional_ pseudo-class, and the parser still need to be extended to enable users (like Servo) to add more of those and implements the parsing of their arguments. I’ll take care of this in a bit.

**Pull request closed without merging**